File: Defender_Scanner.py
# var mpcmdrun = new ProcessStartInfo(@"C:\Program Files\Windows Defender\MpCmdRun.exe")
#         {
#             Arguments=$"-Scan -ScanType 3 -File \"{file}\" -DisableRemediation -Trace -Level 0x10",
#             CreateNoWindow = true,
#             ErrorDialog = false,
#             UseShellExecute = false,
#             RedirectStandardOutput = true,
#             WindowStyle = ProcessWindowStyle.Hidden
#         };
# public enum ScanResult
# {
#     [Description("No threat found")]
#     NoThreatFound,
#     [Description("Threat found")]
#     ThreatFound,
#     [Description("The file could not be found")]
#     FileNotFound,
#     [Description("Timeout")]
#     Timeout,
#     [Description("Error")]
#     Error
# }
import re
import subprocess
import os

def denfender_scan(file):
    if not os.path.exists(file):
        raise Exception(f'No such file: {file}')
    
    exe = "C:\\Program Files\\Windows Defender\\MpCmdRun.exe"
    # command_to_execute = exe + " -Scan -ScanType 3 -File '{file}' -DisableRemediation -Trace -Level 0x10"
    # run = subprocess.run(command_to_execute, capture_output=True)


    mpcmdrun = subprocess.Popen(
    [
        exe,
        "-Scan",
        "-ScanType",
        "3",
        f"-File",
        '"'+file+'"',
        "-DisableRemediation",
        "-Trace",
        "-Level",
        "0x10",
    ],
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE,
    text=True,
    )
    stdout, stderr = mpcmdrun.communicate()
    if stdout:
        stdout = stdout.lower()

        if "no threats" in stdout:
             print("The file '" +file+"' is clean")

        else:
            threat = re.findall('\sthreat\s+:\s+(.*)', stdout)[0]
            #print(stdout)
            print('Threat: '+threat)

    if stderr:
        print("Standard Error:")
        print(stderr)

denfender_scan('F:\\C2\\test-sample\\Invoke-Mimikatz.ps1')
denfender_scan('F:\\C2\\test-sample\\good_file.ps1')

File: ThreatCheck/ThreatCheck/Scanners/DefenderScanner.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;

namespace ThreatCheck.Scanners;

internal sealed class DefenderScanner : Scanner
{
    private readonly byte[] _fileBytes;
    private string _filePath;

    public DefenderScanner(byte[] file)
    {
        _fileBytes = file;
    }

    public void AnalyzeFile()
    {
        if (!Directory.Exists(@"C:\Temp"))
        {
#if DEBUG
            CustomConsole.WriteDebug(@"C:\Temp doesn't exist. Creating it...");
#endif
            Directory.CreateDirectory(@"C:\Temp");
        }

        _filePath = Path.Combine(@"C:\Temp", "file.exe");
        File.WriteAllBytes(_filePath, _fileBytes);

        var status = ScanFile(_filePath);

        if (status.Result is ScanResult.NoThreatFound)
        {
            CustomConsole.WriteOutput("No threat found!");
            return;
        }

        Malicious = true;

        CustomConsole.WriteOutput($"Target file size: {_fileBytes.Length} bytes");
        CustomConsole.WriteOutput("Analyzing...");

        var splitArray = new byte[_fileBytes.Length / 2];
        Buffer.BlockCopy(_fileBytes, 0, splitArray, 0, _fileBytes.Length / 2);
        var lastgood = 0;

        while (!Complete)
        {
#if DEBUG
            CustomConsole.WriteDebug($"Testing {splitArray.Length} bytes");
#endif
            File.WriteAllBytes(_filePath, splitArray);
            status = ScanFile(_filePath);

            if (status.Result is ScanResult.ThreatFound)
            {
#if DEBUG
                CustomConsole.WriteDebug("Threat found, splitting");
#endif
                var tmpArray = HalfSplitter(splitArray, lastgood);
                Array.Resize(ref splitArray, tmpArray.Length);
                Array.Copy(tmpArray, splitArray, tmpArray.Length);
            }
            else
            {
#if DEBUG
                CustomConsole.WriteDebug("No threat found, increasing size");
#endif
                lastgood = splitArray.Length;
                var tmpArray = Overshot(_fileBytes, splitArray.Length);
                Array.Resize(ref splitArray, tmpArray.Length);
                Buffer.BlockCopy(tmpArray, 0, splitArray, 0, tmpArray.Length);
            }
        }
    }

    private static DefenderScanResult ScanFile(string file, bool getSig = false)
    {
        var result = new DefenderScanResult();

        if (!File.Exists(file))
        {
            result.Result = ScanResult.FileNotFound;
            return result;
        }

        using var process = new Process();
        var mpcmdrun = new ProcessStartInfo(@"C:\Program Files\Windows Defender\MpCmdRun.exe")
        {
            Arguments = $"-Scan -ScanType 3 -File \"{file}\" -DisableRemediation -Trace -Level 0x10",
            CreateNoWindow = true,
            ErrorDialog = false,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            WindowStyle = ProcessWindowStyle.Hidden
        };

        process.StartInfo = mpcmdrun;
        process.Start();
        process.WaitForExit(30000); // Wait 30s

        if (!process.HasExited)
        {
            process.Kill();
            result.Result = ScanResult.Timeout;
            return result;
        }

        if (getSig)
        {
            while (process.StandardOutput.ReadLine() is { } stdout)
            {
                if (!stdout.Contains("Threat  "))
                    continue;
                
                var sig = stdout.Split(' ');
                var sigName = sig[19];
                result.Signature = sigName;
                break;
            }
        }

        result.Result = process.ExitCode switch
        {
            0 => ScanResult.NoThreatFound,
            2 => ScanResult.ThreatFound,
            _ => ScanResult.Error
        };

        return result;
    }
}

public class DefenderScanResult
{
    public ScanResult Result { get; set; }
    public string Signature { get; set; }
}

public enum ScanResult
{
    [Description("No threat found")]
    NoThreatFound,
    [Description("Threat found")]
    ThreatFound,
    [Description("The file could not be found")]
    FileNotFound,
    [Description("Timeout")]
    Timeout,
    [Description("Error")]
    Error
}